pkg/controlplane/controller/leaderelection/leaderelection_controller.go

```diff
@@ -0,0 +1,428 @@
+	}
+
+	for _, candidate := range candidates {
+		if candidate.Spec.RenewTime != nil && candidate.Spec.RenewTime.Add(leaseCandidateValidDuration/2).Before(time.Now()) {
+			return true, nil
+		}
+	}
+
+	prelimStrategy, err := pickBestStrategy(candidates)
+	if err != nil {
+		return false, err
+	}
+	if prelimStrategy != v1.OldestEmulationVersion {
+		klog.V(5).Infof("Strategy %q is ignored by CLE", prelimStrategy)
+		return false, nil
+	}
+
+	prelimElectee := pickBestLeaderOldestEmulationVersion(candidates)
+	if prelimElectee == nil {
+		return false, nil
+	} else if lease != nil && lease.Spec.HolderIdentity != nil && prelimElectee.Name == *lease.Spec.HolderIdentity {
+		klog.V(5).Infof("Leader %s is already most optimal for lease %s", prelimElectee.Name, leaseNN)
+		return false, nil
+	}
+	return true, nil
+}
+
+// reconcileElectionStep steps through a step in an election.
+// A step looks at the current state of Lease and LeaseCandidates and takes one of the following action
+// - do nothing (because leader is already optimal or still waiting for an event)
+// - request ack from candidates (update LeaseCandidate PingTime)
+// - finds the most optimal candidate and elect (update the Lease object)
+// Instead of keeping a map and lock on election, the state is
+// calculated every time by looking at the lease, and set of available candidates.
+// PingTime + electionDuration > time.Now: We just asked all candidates to ack and are still waiting for response
+// PingTime + electionDuration < time.Now: Candidate has not responded within the appropriate PingTime. Continue the election.
+// RenewTime + 5 seconds > time.Now: All candidates acked in the last 5 seconds, continue the election.
+func (c *Controller) reconcileElectionStep(ctx context.Context, leaseNN types.NamespacedName) (requeue time.Duration, err error) {
+	now := time.Now()
+
+	candidates, err := c.listAdmissableCandidates(leaseNN)
+	if err != nil {
+		return defaultRequeueInterval, err
+	} else if len(candidates) == 0 {
+		return noRequeue, nil
+	}
+	klog.V(6).Infof("Reconciling election for %s, candidates: %d", leaseNN, len(candidates))
+
+	// Check if an election is really needed by looking at the current lease and candidates
+	needElection, err := c.electionNeeded(candidates, leaseNN)
+	if !needElection {
+		return noRequeue, err
+	}
+	if err != nil {
+		return defaultRequeueInterval, err
+	}
+
+	canVoteYet := true
+	for _, candidate := range candidates {
+		if candidate.Spec.PingTime != nil && candidate.Spec.PingTime.Add(electionDuration).After(now) {
+			// continue waiting for the election to timeout
+			canVoteYet = false
+			continue
+		}
+		if candidate.Spec.RenewTime.Add(electionDuration).After(now) {
+			continue
+		}
+
+		if candidate.Spec.PingTime == nil ||
+			// If PingTime is outdated, send another PingTime only if it already acked the first one.
+			(candidate.Spec.PingTime.Add(electionDuration).Before(now) && candidate.Spec.PingTime.Before(candidate.Spec.RenewTime)) {
```

Comment:
why won't this continuously ping every other candidate if one of them keeps the controller waiting (canVoteYet=false).

Reply:
Added a todo to address in a follow up. It could happen in the rare case that the Update operation takes a very long time to complete. Proposal is to update in parallel, randomize the order, and using the same PingTime.